### PR TITLE
generate templates from version_info

### PR DESCRIPTION
**Pull request by ibrokethecloud**
PR introduces the following:
* uses the version_info file as source of truth to generate a template file
* uses generated template file to generate individual addon manifests

This ensures users only ever need to update the version_info file to ensure templates and addon manifests are generated from single source of truth.

### CODE REVIEW

**Comment by w13915984028** [comment]
@ibrokethecloud 
This PR will further consolidate the data consistency. A few addtional questions:

(1) Is there another PR to run `go run . -generateTemplates -path $path_to_installer_templates` on installer build-bundle?

(2) For rancher-monitoring and rancher-logging, they are a bit complex for history code, the ENV are sourced from version* files under installer and used more;

We can do it via:
(2.1) Add those to version_info
```
LOGGING_VERSION="103.0.0+up3.17.10"
MONITORING_VERSION="103.0.3+up45.31.1"
HARVESTER_EVENTROUTER_IMAGE="rancher/harvester-eventrouter:v0.2.0"
```
(2.2) Refer those 3 in the `pkg/templates/rancherd-22-addons.yaml`

(2.3) In Installer, allow replace env,  addon to replace installer, or installer to replace addon, I am considering which is better.

(3) The images are stored on files like https://github.com/harvester/harvester-installer/blob/master/scripts/images/harvester-additional-images.txt;  not sure all addon images are included to be packed into ISO; this is a place where mis-matching may happen. thanks.

File: pkg/render/addon.go [hunk 1]
```diff
@@ -67,15 +91,71 @@ func Addon(templateSource, destPath string) error {
...
+	var lines []string
+	for versionScanner.Scan() {
+		lines = append(lines, versionScanner.Text())
+	}
+
+	result := make(map[string]string, len(lines))
+	for _, v := range lines {
+		fields := strings.Split(v, "=")
+		if len(fields) == 2 { // need check to ignore hash=bang directive in version info file
+			if strings.Contains(fields[0], imageSuffix) {
```
Comment (w13915984028) [inline pkg/render/addon.go:127]: This way, the `- image: rancher/harvester-eventrouter:v0.2.0` will be a bit tricky

or we just define the right infomation in the version_info
e.g.:
HARVESTER_SEEDER_IMAGE_TAG="v0.3"
HARVESTER_EVENTROUTER_IMAGE="rancher/harvester-eventrouter:v0.2.0"

The render just replace the var with env, no special processing about the `imageSuffix`.

Anohter solution is just to skip the `IMAGE` suffix for `harvester-eventrouter`

**Review by w13915984028 — COMMENTED** [review]
Another question about the image tag, thanks.

File: pkg/render/addon.go (hunk 1 above)
Comment (w13915984028) [reply pkg/render/addon.go:127]: seems only the second solution to skip `image` suffix, the `HARVESTER_SEEDER_IMAGE` is directly used in build-bundle

```
  # seeder images
  echo ${HARVESTER_SEEDER_IMAGE} >> ${image_list_file}
  ```

**Comment by ibrokethecloud** [comment]
yes we will be changing the installer to use generate templates. 

there is also no easy way to handle monitoring addons since they are lot more complicated due to number of moving parts.

@bk201 i wonder if we should repackage the rancher monitoring chart and host it in harvester charts to ensure install upgrade path involves just upgrading the chart.

File: pkg/render/addon.go (hunk 1 above)
Comment (ibrokethecloud) [reply pkg/render/addon.go:127]: we can update build-bundle as well to bring it in sync with what the addon definition will be, there is no hard rule that the variable name cannot be changed 😄

File: pkg/render/addon.go [hunk 2]
```diff
@@ -67,15 +91,71 @@ func Addon(templateSource, destPath string) error {
...
+				// in case of image names we only need tag info and not full image name
+				images := strings.Split(fields[1], ":")
+				result[fields[0]] = strings.Trim(images[1], "\"")
+
+			} else {
+				result[fields[0]] = fields[1]
+			}
+		}
```
[the remark this answers is not in the document]
Comment (ibrokethecloud) [reply pkg/render/addon.go:138]: the lines in this case will be chart versions and not images, the extra split is needed only to handle images differently. I have added a comment to highlight the same.

**Review by w13915984028 — APPROVED** [review]
LGTM, thanks.

**Review by bk201 — APPROVED** [review]

**Comment by bk201** [comment]
@mergifyio backport v1.4

File: pkg/render/addon.go [hunk 3]
```diff
@@ -67,15 +91,74 @@ func Addon(templateSource, destPath string) error {
...
+	tmpl, err := template.New("").Delims("<<", ">>").Parse(string(contents))
+	if err != nil {
+		return err
+	}
+	renderedContent, err := renderTemplate(tmpl, versionFilePath)
+	if err != nil {
+		return fmt.Errorf("error rendering template: %v", err)
+	}
+
+	err = os.WriteFile(filepath.Join(destPath, defaultFileName), renderedContent, 0755)
```
Comment (tserong) [inline pkg/render/addon.go:158]: Should that be 0644?

**Review by tserong — APPROVED** [review]
LGTM - I just noticed one little thing.

File: pkg/render/addon.go (hunk 3 above)
Comment (ibrokethecloud) [reply pkg/render/addon.go:158]: yeah i can change it :D

**Pull request merged**